server/routes.go

```diff
@@ -1554,37 +1557,32 @@ func (s *Server) ChatHandler(c *gin.Context) {
 				res.LoadDuration = checkpointLoaded.Sub(checkpointStart)
 			}
 
-			// TODO: tool call checking and filtering should be moved outside of this callback once streaming
-			// however this was a simple change for now without reworking streaming logic of this (and other)
-			// handlers
-			if req.Stream != nil && !*req.Stream || len(req.Tools) == 0 {
-				ch <- res
-				return
-			}
-
-			// Streaming tool calls:
-			// If tools are recognized, use a flag to track the sending of a tool downstream
-			// This ensures that content is cleared from the message on the last chunk sent
-			sb.WriteString(r.Content)
-			if toolCalls, ok := m.parseToolCalls(sb.String()); ok {
-				res.Message.ToolCalls = toolCalls
-				for i := range toolCalls {
-					toolCalls[i].Function.Index = toolCallIndex
-					toolCallIndex++
+			if len(req.Tools) > 0 && !toolParser.Done {
+				toolCalls, leftover := toolParser.ParseToolCalls(r.Content)
+				// * This can be abstracted again to a .handleState(tp.state)
+				// * However, we'd need a flag to indicate whether to send the response or not
+				// * happy to take whatever is more idiomatic
+				switch toolParser.ParserState {
+				case ToolCallAccumulate:
+					// tokens are accumulated in the tool parser
+					return
+				case ToolCallSendTokens:
+					// tokens are sent back in the response
+				case ToolCallSendPartial:
+					// tokens not needed for parsing are sent back in the response
+					if len(leftover) > 0 {
+						res.Message.Content = leftover
+					}
+					// ! state is needed as we need to not match on the other states
+				case ToolCallFound:
+					res.Message.ToolCalls = toolCalls
+					res.Message.Content = ""
 				}
-				res.Message.Content = ""
-				sb.Reset()
-				ch <- res
-				return
 			}
 
-			if r.Done {
-				// Send any remaining content if no tool calls were detected
-				if toolCallIndex == 0 {
-					res.Message.Content = sb.String()
-				}
-				ch <- res
-			}
+			fmt.Println("sending response", res.Message.Content)
+			// * this is where we'd need the flag if we have a .handleState(tp.state)
+			ch <- res
```

Comment:
Would like your thoughts on what is idiomatic @jmorganca 

Personally, i think this is clearest to understand high level without over abstracting to be direct.